File: Python/snake.py
from node import Node


class Snake(Node):
    def __init__(self, x: int, y: int, next: Node | None) -> None:
        super().__init__(x, y, next)

    def move(self, x: int, y: int, grow: bool) -> bool:
        old_head: Node = Node(self.x, self.y, self.next)
        self.next = old_head

        self.x = x
        self.y = y

        if not grow and self.next is not None:
            if self.next.next is None:
                self.next = None
            else:
                curr: Node = self.next

                while curr.next is not None and curr.next.next is not None:
                    curr = curr.next

                curr.next = None

        return not self._self_collision() and not self._wall_collision()

    def _wall_collision(self) -> bool:
        return self.x < 0 or self.x >= 20 or self.y < 0 or self.y >= 20

    def _self_collision(self) -> bool:
        curr: Node | None = self.next

        while curr is not None:
            if curr.x == self.x and curr.y == self.y:
                return True
            curr = curr.next

        return False

File: Typescript/src/snake.ts
import Node from "./node.js";

export default class Snake {
    public head: Node;

    constructor(x: number, y: number, next: Node) {
        this.head = new Node(x, y, next);
    }

    public move(x: number, y: number, grow: boolean): boolean {
        const old_head = new Node(this.head.x, this.head.y, this.head.next);
        this.head.next = old_head;

        this.head.x = x;
        this.head.y = y;

        if (!grow && this.head.next != null) {
            let curr = this.head;

            while (curr.next != null && curr.next.next != null) {
                curr = curr.next;
            }
            curr.next = null;
        }
        return !this.wallCollision() && !this.selfCollision();
    }

    private wallCollision(): boolean {
        const x = this.head.x;
        const y = this.head.y;

        return x < 0 || x >= 20 || y < 0 || y >= 20;
    }

    private selfCollision(): boolean {
        for (const node of this.head) {
            if (node == this.head) continue;

            if (node.x == this.head.x && node.y == this.head.y) {
                return true;
            }
        }
        return false;
    }
}
